File: Python/Concurrency/RaceDemo.py
import threading
from concurrent.futures import ThreadPoolExecutor


class SafeCounter:
    def __init__(self) -> None:
        self._value = 0
        self._lock = threading.Lock()

    def increment(self) -> None:
        with self._lock:
            self._value += 1

    def get(self) -> int:
        return self._value


unsafe_counter = 0
safe_counter = SafeCounter()


def increment_both() -> None:
    global unsafe_counter
    unsafe_counter += 1  # NOT atomic — read-modify-write across threads
    safe_counter.increment()  # atomic (guarded by a lock)


def main() -> None:
    with ThreadPoolExecutor(max_workers=8) as pool:
        for _ in range(1000):
            pool.submit(increment_both)

    print(f"unsafe: {unsafe_counter}")  # almost never 1000
    print(f"safe:   {safe_counter.get()}")  # always 1000


if __name__ == "__main__":
    main()

File: Java/Concurrency/RaceDemo.java
package Concurrency;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

public class RaceDemo {
    static int unsafeCounter = 0;
    static AtomicInteger safeCounter = new AtomicInteger(0);

    public static void main(String[] args) throws InterruptedException {
        ExecutorService pool = Executors.newFixedThreadPool(8);
        for (int i = 0; i < 1000; i++) {
            pool.submit(
                () -> {
                unsafeCounter++;              // NOT atomic — three steps
                safeCounter.incrementAndGet(); // atomic
            });
        }
        pool.shutdown();
        pool.awaitTermination(5, TimeUnit.SECONDS);
        System.out.println("unsafe: " + unsafeCounter);  // almost never 1000
        System.out.println("safe:   " + safeCounter.get()); // always 1000
    }
}
